pages/workspaceLogs/workspaceLogs.js

```diff
@@ -0,0 +1,168 @@
+// @ts-check
+const express = require('express');
+const router = express.Router();
+const asyncHandler = require('express-async-handler');
+const AWS = require('aws-sdk');
+const z = require('zod');
+
+const config = require('../../lib/config');
+const error = require('../../prairielib/lib/error');
+const sqldb = require('../../prairielib/lib/sql-db');
+const sqlLoader = require('../../prairielib/lib/sql-loader');
+
+const { WorkspaceLogs, WorkspaceVersionLogs } = require('./workspaceLogs.html');
+
+const sql = sqlLoader.loadSqlEquiv(__filename);
+
+/**
+ * Loads logs for the given workspace version from S3. If logs cannot be found,
+ * returns null.
+ *
+ * @param {string | number} workspaceId
+ * @param {string | number} version
+ * @param {string} [startAfter]
+ *
+ * @returns {Promise<{ logs: string, startAfter: string} | null>}
+ */
+async function loadLogParts(workspaceId, version, startAfter) {
+  // TODO: handle errors and missing data.
+  const s3Client = new AWS.S3({ maxRetries: 3 });
+  const logItems = await s3Client
+    .listObjectsV2({
+      Bucket: config.workspaceLogsS3Bucket,
+      Prefix: `${workspaceId}/${version}/`,
+      MaxKeys: 10,
+      StartAfter: startAfter,
+    })
+    .promise();
+
+  if (logItems.Contents.length === 0) {
+    return {
+      logs: '',
+      // If we couldn't load any more logs, reuse the same startAfter key.
+      startAfter,
+    };
+  }
+
+  const logParts = await Promise.all(
```

Comment:
This will load up to 10 logs in parallel - should we limit the concurrency more?

Aside: I'm realizing that while "only include up to a minute's worth of logs in one object" is great for writing, it's suboptimal for reading, as to fully load the logs you'd have to make potentially hundreds of `GET` requests to reassemble the full logs. That's what makes this whole async-loading thing necessary.

I'm not sure how much better we can make this. If S3 had a way to append to files, this would all be really easy, but it doesn't. I think our options are:

- Lots of small writes (current implementation)
- One file that we periodically overwrite with the entire log history (a lot of wasted bandwidth)
- One big write as a container is shutting down (makes real-time debugging impossible)
- One big write as a container is shutting down, and *also* a way to request logs directly from live containers (API to workspace hosts on the fly)

The last option might actually be the best of both worlds - thoughts?

Reply:
I made that change in the most recent few commits (https://github.com/PrairieLearn/PrairieLearn/pull/6638/files/8c73f53f90d30db4e25c1abc179276971f4ca4ee..2de19e9cde7eb9b2be54263047ab20f8e9dc6406) - it feels overall much simpler and more robust.

Reply:
Agreed, this new code is a definite improvement.